# Homework Assignment #4 (Individual)
## Regression and SVM Predictive Models

### <p style="text-align: right;"> &#9989; Ritik Jain.</p>
### <p style="text-align: right;"> &#9989; Ritikjain1.</p>

### Goal for this homework assignment
By now, you have learned a bit about predictive models. In this assignment, you will practice:

* Building predictive models using `statsmodels` and `sklearn`
* Evaluating your predictive models

**This assignment is due roughly two weeks from now at 11:59 pm on Friday, April 21st.** It should be uploaded into the "Homework Assignments" submission folder for Homework #4.  Submission instructions can be found at the end of the notebook.


---
## Part 0: Add to your Git repository to track your progress on your assignment (4 points)

For this assignment, you're going to add it to the `cmse202-s23-turnin` repository you created in class so that you can track your progress on the assignment and preserve the final version that you turn in. In order to do this you need to

**&#9989; Do the following**:

1. Navigate to your `cmse202-s23-turnin` repository and create a new directory called `hw-04`.
2. Move this notebook into that **new directory** in your repository, then **add it and commit it to your repository**.
1. Finally, to test that everything is working, "git push" the file so that it ends up in your GitHub repository.

**Important**: Double check you've added your Professor and your TA as collaborators to your "turnin" repository (you should have done this in the previous homework assignment).

**Also important**: Make sure that the version of this notebook that you are working on is the same one that you just added to your repository! If you are working on a different copy of the notebook, **none of your changes will be tracked**!

If everything went as intended, the file should now show up on your GitHub account in the "`cmse202-s23-turnin`" repository inside the `hw-04` directory that you just created.  Periodically, **you'll be asked to commit your changes to the repository and push them to the remote GitHub location**. Of course, you can always commit your changes more often than that, if you wish.  It can be good to get into a habit of committing your changes any time you make a significant modification, or when you stop working on the project for a bit.

&#9989; **Do this**: Before you move on, put the command that your instructor should run to clone your repository in the markdown cell below.

# Building a Model to Predict if it Will Rain Tomorrow

In this dataset we have weather data for a number of Australian cities. We’re going to **build a model that uses data from today to predict if it will rain tomorrow.** To simplify our model, let’s just look at the weather for one city.

## Part 1. Working with the Data

The dataset that we’ll be using can be found here:
- `https://raw.githubusercontent.com/msu-cmse-courses/cmse202-S23-data/main/HW/HW4/aussie_weather_data.csv`

#### 1.1 (3 Points)
**Download the data set and read it in using Pandas.**

In [1]:
#Write your code here
import pandas as pd
Weather = pd.read_csv('aussie_weather_data.csv')
Weather

Year  Month  DayOfMonth       Location  MinTemp  MaxTemp  Rainfall  \
0      2009      1           1  NorfolkIsland     20.4     25.8       0.0   
1      2009      1           2  NorfolkIsland     20.9     26.7       0.2   
2      2009      1           3  NorfolkIsland     22.3     26.3       0.0   
3      2009      1           4  NorfolkIsland     21.6     22.2       1.2   
4      2009      1           5  NorfolkIsland     20.4     23.5       2.6   
...     ...    ...         ...            ...      ...      ...       ...   
39510  2017      6          20         Darwin     19.3     33.4       0.0   
39511  2017      6          21         Darwin     21.2     32.6       0.0   
39512  2017      6          22         Darwin     20.7     32.8       0.0   
39513  2017      6          23         Darwin     19.5     31.8       0.0   
39514  2017      6          24         Darwin     20.2     31.7       0.0   

       Evaporation  Sunshine  WindGustDir  ...  Humidity9am  Humidity3pm  \
0              6.0      12.4        112.5  ...         84.0         71.0   
1              8.0      10.3         90.0  ...         79.0         77.0   
2              3.2       2.0          0.0  ...         87.0         90.0   
3              2.8       0.0        337.5  ...         92.0         95.0   
4              2.2       2.9        337.5  ...         86.0         86.0   
...            ...       ...          ...  ...          ...          ...   
39510          6.0      11.0         22.5  ...         63.0         32.0   
39511          7.6       8.6          0.0  ...         56.0         28.0   
39512          5.6      11.0          0.0  ...         46.0         23.0   
39513          6.2      10.6        337.5  ...         62.0         58.0   
39514          5.6      10.7         22.5  ...         73.0         32.0   

       Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0           1014.5       1013.6       3.0       1.0     23.3     24.7   
1           1016.3       1015.5       2.0       5.0     25.0     25.1   
2           1014.6       1014.3       7.0       7.0     24.7     23.8   
3           1016.0       1015.3       8.0       8.0     22.1     21.2   
4           1015.3       1013.7       7.0       7.0     21.8     21.6   
...            ...          ...       ...       ...      ...      ...   
39510       1013.9       1010.5       0.0       1.0     24.5     32.3   
39511       1014.6       1011.2       7.0       0.0     24.8     32.0   
39512       1015.3       1011.8       0.0       0.0     24.8     32.1   
39513       1014.9       1010.7       1.0       1.0     24.8     29.2   
39514       1013.9       1009.7       6.0       5.0     25.4     31.0   

       RainToday  RainTomorrow  
0              0             0  
1              0             0  
2              0             1  
3              1             1  
4              1             0  
...          ...           ...  
39510          0             0  
39511          0             0  
39512          0             0  
39513          0             0  
39514          0             0  

[39515 rows x 25 columns]

#### 1.2 (3 Points)
This dataset has weather information for multiple cities in Australia. It’s quite a large dataset. In fact, it’s a little *too* large for our purposes; it will take a considerable amount of time to train a model on so much data. So we’ll just work with the data for a specific city.

The first thing we’ll do is see which cities we have data for. 
**Print out a list of the specific cities in this dataset.** (Note that this is not the same as printing the entire `Location` column.)

In [2]:
#Write your code here
Weather['Location'].unique()

array(['NorfolkIsland', 'SydneyAirport', 'WaggaWagga', 'MelbourneAirport',
       'Mildura', 'Watsonia', 'Brisbane', 'Cairns', 'Townsville',
       'MountGambier', 'Nuriootpa', 'PerthAirport', 'Perth',
       'AliceSprings', 'Darwin'], dtype=object)

#### 1.3 (3 Points)
Select one of the cities (and **please** don't just pick the first one!). Cut down your dataframe so that it only contains data from your chosen city. You should also **look through the columns in this dataset and ensure that all of the data is ready for our analysis/model creation.**

In [3]:
#Write your code here
Brisbane_Weather = Weather[Weather['Location'] == 'Brisbane']
Brisbane_Weather

Year  Month  DayOfMonth  Location  MinTemp  MaxTemp  Rainfall  \
16003  2008      7           1  Brisbane     10.3     25.2       0.0   
16004  2008      7           2  Brisbane      8.1     22.9       0.0   
16005  2008      7           3  Brisbane      9.7     22.4       0.0   
16006  2008      7           4  Brisbane     11.8     20.0       0.8   
16007  2008      7           5  Brisbane     12.3     16.7       0.0   
...     ...    ...         ...       ...      ...      ...       ...   
18951  2017      6          21  Brisbane     12.2     23.4       0.0   
18952  2017      6          22  Brisbane     13.7     22.7       0.0   
18953  2017      6          23  Brisbane     10.3     23.7       0.0   
18954  2017      6          24  Brisbane     10.4     24.5       0.0   
18955  2017      6          25  Brisbane     11.0     24.2       0.0   

       Evaporation  Sunshine  WindGustDir  ...  Humidity9am  Humidity3pm  \
16003          1.4       9.5        157.5  ...         81.0         37.0   
16004          2.0       9.8        180.0  ...         41.0         30.0   
16005          5.8       9.4          0.0  ...         55.0         52.0   
16006          1.8       1.1        225.0  ...         76.0         53.0   
16007          2.0       0.3        270.0  ...         81.0         89.0   
...            ...       ...          ...  ...          ...          ...   
18951          4.0       9.2        202.5  ...         61.0         44.0   
18952          3.4       7.2        315.0  ...         62.0         55.0   
18953          3.2       9.4        180.0  ...         71.0         37.0   
18954          3.4       8.9        270.0  ...         75.0         33.0   
18955          2.2       9.8         22.5  ...         68.0         53.0   

       Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
16003       1019.6       1014.8       0.0       1.0     14.9     24.6   
16004       1018.8       1015.0       0.0       0.0     16.2     22.4   
16005       1021.4       1019.1       1.0       4.0     15.4     21.3   
16006       1023.5       1021.7       7.0       7.0     14.1     19.6   
16007       1027.3       1026.2       7.0       8.0     16.1     15.0   
...            ...          ...       ...       ...      ...      ...   
18951       1023.4       1021.3       1.0       4.0     17.2     23.3   
18952       1026.0       1022.4       2.0       3.0     17.8     20.5   
18953       1023.3       1018.4       1.0       3.0     15.2     23.1   
18954       1018.6       1015.4       7.0       5.0     14.3     24.0   
18955       1020.5       1017.3       6.0       3.0     15.9     22.6   

       RainToday  RainTomorrow  
16003          0             0  
16004          0             0  
16005          0             0  
16006          0             0  
16007          0             1  
...          ...           ...  
18951          0             0  
18952          0             0  
18953          0             0  
18954          0             0  
18955          0             0  

[2953 rows x 25 columns]

#### 1.4 (3 Points)
Finally, before we start creating our models, let's first split our data into training and testing datasets. **Keep in mind what we want our model to predict and how this dataset gives you the information you need for your model.**

In [4]:
#Write your code here
from sklearn.model_selection import train_test_split

X = Brisbane_Weather.drop(columns=['MaxTemp','Year','Month','DayOfMonth','Location'], axis = 1)
y = Brisbane_Weather['MaxTemp']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## 2.0 Building a Regression Model

#### 2.1 (3 Points)
We’ll start by creating a regression model. **Does it make sense for us to use linear regression or logistic regression? Explain your choice.**

Linear regression because it is easier to use linerar regression to find the relationship between max temp and the weather features. logistical regression doesnt make sense here because it is used for binary classification problems, and this is not one of them.

#### 2.2 (12 Points)
Create a regression model. Use the model summary to determine which parameters are most statistically significant in your model. If you get an error during at this stage it may be helpful to go back to part 1.3 and ensure that all of your data is appropriate for creating a model. 

In [5]:
#Write your code here
import statsmodels.api as sm
from sklearn.linear_model import LinearRegression

X_train = sm.add_constant(X_train)
ols_model = sm.OLS(y_train, X_train).fit()
print(ols_model.summary())

lr_model = LinearRegression()
lr_model.fit(X_train, y_train)
print("R^2", lr_model.score(X_train, y_train))

                            OLS Regression Results                            
Dep. Variable:                MaxTemp   R-squared:                       0.955
Model:                            OLS   Adj. R-squared:                  0.955
Method:                 Least Squares   F-statistic:                     2484.
Date:                Fri, 21 Apr 2023   Prob (F-statistic):               0.00
Time:                        17:50:34   Log-Likelihood:                -2885.6
No. Observations:                2362   AIC:                             5813.
Df Residuals:                    2341   BIC:                             5934.
Df Model:                          20                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const            15.2234      4.258      3.575

*all the numbers in the column are pretty small with the highest being .849 and the regression score being .95 which is pretty close to 1, which is good.*

#### 2.3 (8 Points)
**Fit your test data and create/print a confusion matrix. We’ll use this to evaluate how well your predicitve model performs.**

In [7]:
#Write your code here
from sklearn.metrics import mean_absolute_error, mean_squared_error
import numpy as np

X_test = sm.add_constant(X_test)
test_score = lr_model.score(X_test, y_test)

print("R^2:", test_score)

y_pred = ols_model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)

print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)

R^2: 0.942241743682302
MAE: 0.6909386937229283
MSE: 0.8804843309310438
RMSE: 0.9383412657082943


#### 2.4 (3 Points)
Look at/calculate the Precision and Recall of your model. **Which one is greater? Describe in plain language what that means about the performance of your model (I.e., the circumstances in which it does/doesn’t do well).**

*In a linear regression model we use the R^2 model, and it is .94 which is pretty good because it is close to one so the preformence of the model is good, could be better.*

## 3.0 Interlude: A Base Rate Frequency Model (12 Points)

If we didn't create a fancy model for predicting future rainfall, would that mean that we had *no way* of predicting if it would rain? Of course not. There's always *some* kind of model we can use (they just won't be very sophisticated). 

In the case of our problem--predicting whether it will rain tomorrow--what is the simplest model we could use? We'd need to look at the **base rate frequency**. If we know that, over the course of the year, it rains ~30% of the time, that would be our base rate frequency. If we were to guess that it would be sunny every day of the year, we'd be wrong ~30% of the time, *but we'd be right ~70% of the time.* 

Why is this important? Imagine we create a sophisticated model that accurately predicts rain/no rain ~60% of the time; without any other information, we might think, "That's not too bad, I guess." It certainly feels better than *no* information, right? Now think about the base rate frequency model, which achieves an accuracy of **70%** just by guessing the same thing every time. Suddenly, our sophisticated model seems like crap! It tells us we could be doing ~10% better than our fancy model by guessing. This is why we use the base rate frequency as a simple way to test our models.


**Calculate how frequently it rains in your city. If you were to guess that it *wasn't* going to rain every day, how frequently would you be right? (We'll use this as our baseline model.)**


In [8]:
#Write your code for calculating the frequency of rain here.
rainy_days = len(Brisbane_Weather[Brisbane_Weather['RainTomorrow'] == 1])
total_days = len(Brisbane_Weather)
rain_frequency = rainy_days / total_days
print(rain_frequency)

0.21740602776837115


In [9]:
accuracy = 1 - rain_frequency
print(accuracy)

0.7825939722316289


*The model will be accurate 78% of the time*

---

### &#128721; STOP (3 Point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit the changes to your Git repository with a meaningful commit message.

---

## 4.0 SVM Model

We’ve seen how regression faired in predicting rain; now let’s give Support Vector Machines a shot. 

It’s not clear, *a priori*, which type of kernel (linear or RBF) would fair better. Therefore, we’ll need to test out both kernel types, as well as multiple values for hyperparameters ($C$ and $\gamma$). 

#### 4.1 (12 Points)
Run a grid search over both types of kernels and multiple values of $C$ and $\gamma$ (we recommend [0.01,0.1,1.0,10.0] for both). We *STRONGLY* recommend you set the parameter `n_jobs=-1` in `GridSearchCV`, which will distribute the computational load. Make sure to write down your best-fit hyperparameters. 

In [10]:
from sklearn.preprocessing import LabelBinarizer

lb = LabelBinarizer()
y_train_bin = (y_train.round() > 30).astype(int)

In [11]:
#Write your code here
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
import numpy as np

param_grid = {
    'C': [0.01, 0.1, 1.0, 10.0],
    'gamma': [0.01, 0.1, 1.0, 10.0]
}

svm = SVC()
grid_search = GridSearchCV(estimator=svm, param_grid=param_grid, cv=5, n_jobs=-1)
grid_search.fit(X_train, y_train_bin)
print( grid_search.best_params_)


{'C': 1.0, 'gamma': 0.01}


*c = 1 and gamma = .01*

#### 4.2 (8 Points)
Using your best-fit parameters, classify your test data and print off the confusion matrix. Note/calculate the Precision and Recall for your SVM model, as well as the overall accuracy.

$\mathrm{Accuracy} = \frac{\mathrm{Accurate~Predictions}}{\mathrm{Total~Samples}}$.

In [12]:
#Write your code here
from sklearn.metrics import confusion_matrix

svm_model = SVC(kernel='rbf', C=1.0, gamma=0.1)
svm_model.fit(X_train, y_train_bin)
y_pred = svm_model.predict(X_test)

y_test_bin = (y_test.round() > 30).astype(int)
conf_matrix = confusion_matrix(y_test_bin, y_pred)
print(conf_matrix)

tp = conf_matrix[1,1]
tn = conf_matrix[0,0]
fp = conf_matrix[0,1]
fn = conf_matrix[1,0]

precision = tp / (tp + fp)
recall = tp / (tp + fn)
accuracy = (tp + tn) / (tp + tn + fp + fn)

print('Precision:', precision)
print('Recall:', recall)
print('Accuracy:', accuracy)

[[493   0]
 [ 98   0]]
Precision: nan
Recall: 0.0
Accuracy: 0.8341793570219966


<ipython-input-12-f616fefa110f>:17: RuntimeWarning: invalid value encountered in long_scalars
  precision = tp / (tp + fp)


## 5. Evaluation of Models

Here we’re going to investigate how well our predictive models actually work.


#### 5.1 (5 Points)

Compare the accuracy of both of your predictive models (regression and SVM) to the baseline model from part 3. Do your predictive models do better than the baseline model? Is their performance relative to the baseline model what *you* would expect? 


*both the svm and the regular model is better than the the baseline, the R^2 is better and the RMSE is better than the basline model.*

#### 5.2 (5 Points)

Let’s test the robustness of your model. Before you do anything, you should record the precision, recall, and accuracy of your two predictive models (if you haven’t already). Once you’ve done that, go all the way back to part 1.4 and change the random seed for `train_test_split`. Using this new data split, rerun all of your code from parts 2 and 4. For each new random seed, record the precision, recall, and accuracy for both models (and write them, as well as the random seed, in the cell below!). Do this for 3-5 different random seeds. 


*Precision: nan Recall: 0.0 Accuracy: 0.8341793570219966*

#### 5.3 (5 Points)

How much variation did you observe in the precision, recall, and accuracy in your models? After going through this exercise, how (if at all) does this change your response to part 5.1? 

*the one in the linear regresson model was much better, i may have messed up somewhere with the svc model but from what i saw the regression model was accurate.*

#### 5.4 (5 Points)

Finally, compare the performance of your regression and SVM models. Which, if any, performed better? Esxplain.

*from what in have the regresson worked better because it gave me the score and it was pretty accurate as well.*

---

### &#128721; STOP (3 Point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit the changes to your Git repository with a meaningful commit message.

---

### Congratulations, you're done!

&#169; Copyright 2022,  Department of Computational Mathematics, Science and Engineering at Michigan State University